### PR TITLE
OS-8206 Enable piadm(1M) to detect and install on-disk iPXE on Triton CNs

### PR DESCRIPTION
**Pull request by danmcd**
Stubs, and the ability to detect a Triton CN.  Do not use yet save to test "yes I'm on a CN" vs. "No I'm not on a CN".

### CODE REVIEW

**Comment by danmcd** [comment]
Test run from this morning using the latest piadm per this PR:

```
[root@larry (kebecloud) ~]# /tmp/piadm list
PI STAMP               BOOTABLE FILESYSTEM            BOOT IMAGE NOW  NEXT 
20200930T173623Z       zones/boot                     next       no   no  
ipxe                   zones/boot                     next       no   yes 
[root@larry (kebecloud) ~]# /tmp/piadm -v list
PI STAMP               BOOTABLE FILESYSTEM            BOOT IMAGE NOW  NEXT 
20200930T173623Z       zones/boot                     next       no   no  
ipxe(20200626T133005Z) zones/boot                     next       no   yes 
[root@larry (kebecloud) ~]# zfs list -t all | grep boot
zones/boot                                         600M   284G   304M  /zones/boot
zones/boot@baseline                                297M      -   300M  -
[root@larry (kebecloud) ~]# zfs rollback zones/boot@baseline
[root@larry (kebecloud) ~]# /tmp/piadm -v list
PI STAMP               BOOTABLE FILESYSTEM            BOOT IMAGE NOW  NEXT 
20200910T155139Z       zones/boot                     none       no   no  
ipxe(20200219T112943Z) zones/boot                     next       no   yes 
[root@larry (kebecloud) ~]# 
[root@larry (kebecloud) ~]# /tmp/piadm -v bootable
zones                          ==> BIOS 
[root@larry (kebecloud) ~]# /tmp/piadm -v bootable -e
To enable a pool for booting, please specify at last a pool

Usage: piadm [-v] <command> [command-specific arguments]

    piadm activate|assign <PI-stamp> [ZFS-pool-name]
    piadm avail
    piadm bootable [-d] [-e [-i <source>]] [-r] [ZFS-pool-name]
    piadm install <source> [ZFS-pool-name]
    piadm list <-H> [ZFS-pool-name]
    piadm remove <PI-stamp> [ZFS-pool-name]
    piadm update [ZFS-pool-name]

[root@larry (kebecloud) ~]# /tmp/piadm -v bootable -e zones
Pool zones appears to be bootable.
Use 'piadm update' to update the CN's iPXE and backup PI.
[root@larry (kebecloud) ~]# /tmp/piadm -v update
Selecting lone boot pool zones by default.
Updating backup PI to 20201002T173911Z
Installing as a backup Platform image PI stamp 20201002T173911Z
making platform-20201002T173911Z directories
Pulling unix
Pulling boot_archive
Pulling boot_archive.hash
Pulling boot_archive.manifest
Pulling boot_archive.gitstatus
...success installing 20201002T173911Z
Updating iPXE provided by headnode (ver: 20200626T133005Z)
Updating boot provided by headnode (ver: 20200930T173623Z)
[root@larry (kebecloud) ~]# /tmp/piadm -v update
Selecting lone boot pool zones by default.
No updates needed for iPXE and its loader.
If you think there should be an update, run
'sdcadm experimental update-gz-tools' on your Triton Head Node
[root@larry (kebecloud) ~]# /tmp/piadm update
[root@larry (kebecloud) ~]# echo $?
0
[root@larry (kebecloud) ~]# /tmp/piadm -v list
PI STAMP               BOOTABLE FILESYSTEM            BOOT IMAGE NOW  NEXT 
20201002T173911Z       zones/boot                     none       yes  no  
ipxe(20200626T133005Z) zones/boot                     next       no   yes 
[root@larry (kebecloud) ~]# /tmp/piadm list
PI STAMP               BOOTABLE FILESYSTEM            BOOT IMAGE NOW  NEXT 
20201002T173911Z       zones/boot                     none       yes  no  
ipxe                   zones/boot                     next       no   yes 
[root@larry (kebecloud) ~]# 
```

**Comment by danmcd** [comment]
Prior testing has given me confidence enough to promote this out of draft.

**Comment by bahamat** [comment]
I think this is looking really good. 3 things to note:

1. Triton nodes never used CD for booting
2. KEBE comments
3. shellcheck cleanup

File: src/piadm.sh
Comment (bahamat) [inline src/piadm.sh:697]: Change to:

```
initialize_as_CN() {                                                            
        TRITON_CN="yes"                                                         
                                                                                
        source /lib/sdc/config.sh                                               
        load_sdc_config                                                         
                                                                                
        # Establish the CNAPI default boot Platform Image                       
        # shellcheck disable=2154                                               
        cnapi_domain=$("${CURL[@]}" http://"${CONFIG_sapi_domain}"/applications?name=sdc | json -Ha metadata.cnapi_domain)
        CNAPI_DEFAULT_PI=$("${CURL[@]}" http://"${cnapi_domain}"/boot/default | json platform)
}                                                                               
```

Comment (danmcd) [reply src/piadm.sh]: Oh damn, you edited this.  Lemme try again...

Comment (danmcd) [reply src/piadm.sh]: ... there we go.

File: src/piadm.sh
Comment (bahamat) [inline src/piadm.sh]: I think this actually should be:

    eecho "POSSIBLE CORRUPTION: $*"

File: src/piadm.sh
Comment (bahamat) [inline src/piadm.sh:722]: Cleaner (but not a big deal):

    archive_prefix=$(sed 's#kernel/amd64/unix#amd64#g' <<< "$unix_path")

File: src/piadm.sh
Comment (bahamat) [inline src/piadm.sh:726]: Cleaner:

    grep -q "$1" <<< "$unix_path" || corrupt "PI $1 is not in $unix_path"

File: src/piadm.sh
Comment (bahamat) [inline src/piadm.sh:842]: I don't think this is going to match, because the pattern is `20[0-9]{6}T[0-9]{6}Z`. I.e., you're going to miss the `T`

**Review by bahamat — CHANGES_REQUESTED** [review]

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh]: Will fix.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:722]: Will hold off on this one.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:726]: Same here, holding off.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:842]: Shell globs aren't proper regexes.
```
[root@larry (kebecloud) ~]# pargs $$
128610:	-bash
argv[0]: -bash
[root@larry (kebecloud) ~]# ls -d /zones/boot/platform-[0-9]*Z
/zones/boot/platform-20201002T173911Z
[root@larry (kebecloud) ~]# 
```
it'll match.

Comment (danmcd) [reply src/piadm.sh]: Fixed with platform-[0-9]*T*Z

**Review by bahamat — DISMISSED** [review]

**Comment by danmcd** [comment]
Need to re-run tests (initial batches were quite helpful, thanks @bahamat), verify no regressions, etc. etc., and then request IA.

**Comment by danmcd** [comment]
Darn it, while inspecting the diffs to write the test plan, I found a missing bit.  It will be tested as part of the final batch of tests test-plan.  (Sorry @bahamat )

**Comment by danmcd** [comment]
Last change was inspired by field-testing against missing PIs that were booted (and then deleted by the head node).

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:721]: nit: remove the grep
`bootparams | awk -F= '$1 == "boot-file" { print $2 }'`

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:722]: $1 doens't need to be quoted iirc.

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:781]: You should quote this variable for safety....example:
```
link@skullcanyon:~$ foo="foo bar"
link@skullcanyon:~$ ./args.sh ${foo}
2 args: <foo> <bar>
link@skullcanyon:~$ ./args.sh "${foo}"
1 args: <foo bar>
```

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:787]: Quote "${TRITON_IPXE_BOOT}"

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:848]: `"${#pdirs[@]}"`

From the Bash guide:
`The syntax "${myfiles[@]}" is extremely important. It works just like "$@" does for the positional parameters: it expands to a list of words, with each array element as one word, no matter what it contains. Even if there are spaces, tabs, newlines, quotation marks, or any other kind of characters in one of the array elements, it'll still be passed along as one word to whichever command we're running. `

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:850]: `"${#pdirs[@]}"`

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:873]: Maybe we should be defensive about mktemp failing? Not sure if we are running with `set -e`
`tfile=$(mktemp) || fatal "failed to create temp file"`

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:900]: "`${TRITON_IPXE_ETC}"`

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:904]: "`${TRITON_IPXE_ETC}"`

File: src/piadm.sh
Comment (papertigers) [inline src/piadm.sh:973]: `"$installsource"`

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:698]: Thanks for the safety tip.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:727]: Actually, since I have the patient open again for @papertigers , I'll fix these two as well.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:781]: These are hardcoded and not subject to environment or user-input weirdness.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:787]: Same here.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:848]: I had tested this and it worked.  What are you worried about here exactly?  I don't see a number needing to be quoted here.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:850]: See above

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:900]: See above - it's hardcoded and not environmentally overridden.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:904]: Same here.

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh:973]: Will fix.

**Review by danmcd — COMMENTED** [review]
Some of the please-quote suggestions do not tickle shellcheck already, so I might pass on those (unless both of you are in unison).

Fixed the unnecessary pipe-stages, including one I'd not done from @bahamat 's last review.

Watch for a push.

**Comment by papertigers** [comment]
If @bahamas10 has time I'd defer to him about quoting things vs what shellcheck says. He's the resident bash expert.

File: src/piadm.sh
Comment (bahamat) [reply src/piadm.sh:848]: It doesn’t need to be quoted because it’s not expanding the variable to something that can potentially glob. It’s always going to be an integer (length of the array).

File: src/piadm.sh
Comment (bahamat) [reply src/piadm.sh]: And now that I see it...I have to 1-up @papertigers here.

    unix_path=$(bootparams | awk -F= ‘/^boot-file=/ {print $2}’

I.e., avoid `if` if a regexp works.

File: src/piadm.sh
Comment (papertigers) [reply src/piadm.sh:848]: You know what I completely blew past it was getting the len() of the array.  I missed the `#`.  Ignore me!

File: src/piadm.sh
Comment (danmcd) [reply src/piadm.sh]: Syntax error?

```
[root@larry (kebecloud) ~]# unix_path=$(bootparams | awk -F= ‘/^boot-file=/ {print $2}’)
awk: syntax error at source line 1
 context is
	 >>> ? <<< 
awk: bailing out at source line 1
[root@larry (kebecloud) ~]# 
```

Comment (danmcd) [reply src/piadm.sh]: Old comment deleted because tick vs. backtick was the problem.

File: src/piadm.sh
Comment (papertigers) [reply src/piadm.sh:850]: My bad again

File: src/piadm.sh
Comment (papertigers) [reply src/piadm.sh:873]: I guess if mktemp fails we are probably in more trouble.  Up to you if you want to be defensive here.
If not I can mark as resolved and +1

Comment (danmcd) [reply src/piadm.sh:873]: Honestly, if mktemp fails we have much bigger problems.

**Review by papertigers — APPROVED** [review]

**Pull request merged**
